README.org

```diff
@@ -1112,6 +1112,86 @@ Please provide the necessary important information with your bug report:
 When evaluating Consult-related code snippets you should enable [[https://www.gnu.org/software/emacs/manual/html_node/elisp/Lexical-Binding.html][lexical binding]].
 Consult often relies on lambdas and lexical closures.
 
+* Hacking
+** Creating a new asynchronous completion source
+
+If you have a completion source that's both dynamic and expensive to generate,
+=completing-read= may not be the best choice. Instead, =consult--read= serves as a
+thin wrapper around =completing-read= that provides this functionality.
+
+For example, consider the following script that splits its input on space:
+
+#+begin_src sh
+  #!/usr/bin/env bash
+
+  # simulate work
+  sleep .1
+
+  # generate completion candidates
+  printf "%s\n" "$*" | tr " " "\n" | sort
+#+end_src
+
+Let's assume this script is callable as =testibus hello world=.
+
+To have Consult use it as a completion source, use =consult--async-command=:
+
+#+begin_src emacs-lisp
+  (consult--read
+   (consult--async-command (lambda (input) (list "testibus" (s-trim input))))
+   :prompt "run testibus: ")
+#+end_src
+
+*** ...with live preview
+
+Implementing live preview is a little more involved and requires the definition
+of a =STATE= function as defined by =consult--with-preview=. The
+=consult--define-state= macro can assist by defining a function suitable to
+provide to =consult--read='s =:state= argument.
+
+First, define a 'preview' function with a name using the format
+=consult--%s-preview=. When called with no arguments, it should return a suitable
+'state' function:
+
+#+begin_src emacs-lisp
+  (defun consult--testibus-preview ()
+    (lambda (action candidate)
+      (pcase action
+        ('preview
+         (with-current-buffer-window " *testibus*" nil nil
+           (erase-buffer)
+           (insert (format "input: %s\n" (or candidate "(nothing yet!)"))))))))
+#+end_src
+
+Next, define the 'action' function using the format =consult--%s-action=:
+
+#+begin_src emacs-lisp
+  (defun consult--testibus-action (cand)
+    (message "You chose '%s'!" cand))
+#+end_src
+
+This function is called when a completion candidate is accepted.
+
+We're now ready to define a function for =:state= via the ==consult--define-state=
+macro:
+
+#+begin_src emacs-lisp
+  (consult--define-state testibus)
+#+end_src
+
+Finally, we can use the function generated by =consult--define-state= in
+=consult--read=:
+
+#+begin_src emacs-lisp
+  (consult--read
+   (consult--async-command (lambda (input) (list "testibus" (s-trim input))))
```

Comment:
Could you also provide an example not relying on `consult--async-command`? See `consult--dynamic-collection`. Also an example based directly on the underlying async machinery (without `consult--async-command` and without `consult--dynamic-collection`) would be helpful for understanding.

Reply:
I've added an example with `consult--dynamic-collection`.

As for a more DIY approach – I'm not sure I'm the best person to write that. To be honest, I only started messing with Consult internals yesterday :-) I'm not sure how the async machinery actually works myself.